### PR TITLE
Add the ability to override app configurations at the command line

### LINKED ISSUE #1115: Provide command line option to override app settings
### What is the problem or limitation you are having?

PR #1107 initially failed CI because the verify-app tests require the use of a specific template branch on each repository.

### Describe the solution you'd like

It should be possible to override any app setting at the command line. For example:

```
briefcase create --app-setting template_branch=foobar
```
or
```
briefcase create -S template_branch=foobar
```

would be the equivalent of editing `pyproject.toml` and adding a template_branch setting for the platform being built. 

It should be possible to provide multiple setting overrides:

```
briefcase create -S template=https://example.com/template -S template_branch=foobar
```

Argparse `action="append"` allows for this.

### Describe alternatives you've considered

It is possible to override the template branch by changing the app configuration file. However, it is undesirable to modify a configuration file in a CI context, because the modification will become unnecessary as soon as CI passes.

### Additional context

While the motivating use case for this feature is PRs that require a custom template branch, another use of this feature would be modifying the version at build time. By passing in `--app-setting version=1.2.3`, it would be possible to use CI to control the app version.

PR #1106 adds a `finalize()` method to the base command; the end of this method would seem a reasonable place to apply any overrides. This then guarantees that command line arguments are the final applied values for any app setting.

### PR DESCRIPTION
**Pull request by freakboy3742**
Adds a `-C` / `--config` option that allows a user to override the value of a setting at the command line.

e.g.:
```
briefcase create -C template=\"https:/example.com/template\"
```
would generate an app using the values in the pyproject.toml, except for using a different template.

This is primarily useful for internal testing setups, such as generating a test app inside a template repository, or running CI for the Apple support package in CI. However, it could also be useful to pass in other values, such as a version number.

It's likely only of any practical use in the `create` command, but it's been implemented in all "app facing" commands in case it might be useful.

Fixes #1115.

/cc @rmartin16

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: src/briefcase/commands/base.py
Comment (rmartin16) [inline src/briefcase/commands/base.py:123]: I think Rich's stacktrace handles this regardless but can you raise from the `ValueError`?
```suggestion
            except ValueError as e:
                raise BriefcaseCommandError(
                    f"Unable to parse configuration override {value}"
                ) from e
```

File: src/briefcase/commands/base.py
Comment (rmartin16) [inline src/briefcase/commands/base.py:729]: nitpick: no period at the end to match the other options (although, I also shouldn't have added one to `-v`...)
```suggestion
            help="Override the value of the app configuration item KEY with VALUE",
```

File: src/briefcase/commands/run.py
Comment (rmartin16) [inline src/briefcase/commands/run.py:251]: needs to be passed to `finalize()`?

File: src/briefcase/commands/update.py
Comment (rmartin16) [inline src/briefcase/commands/update.py:71]: needs to be passed to `finalize()`?

**Review by rmartin16 — APPROVED** [review]
Well timed; this is meeting my [needs](https://github.com/beeware/.github/actions/runs/6892256834/job/18749136654#step:19:188) of creating specific projects in CI.
```
briefcase create linux AppImage \
    --config 'template="../../"' \
    --config 'requires=["toga-gtk==0.3.1", "PyGobject<3.46.0"]'
```
I guess I hadn't considered this before but...apparently attributes can have spaces.
```
briefcase package -C '"zzz template"={"asdf"=42}'
```
I gave up trying to get a newline character in to the value for `-C`....but that shouldn't matter since you can pass multiple values for `-C`.

👍🏼

File: src/briefcase/commands/base.py
Comment (mhsmith) [reply src/briefcase/commands/base.py:123]: That should be unnecessary: exceptions raised within an except block are automatically chained unless you say `from None`.

**Comment by mhsmith** [comment]
@rmartin16: I'm guessing from your comments that you didn't intend to approve this yet?

File: src/briefcase/commands/base.py
Comment (mhsmith) [inline src/briefcase/commands/base.py:119]: I can't see any mention of multi-level config values, like `--config 'linux.requires=...'`, which is [valid TOML](https://toml.io/en/v1.0.0#keys).

I think this code might actually work with that as long as you only had one of them. But then if you added `--config linux.something_else=...`, the second `linux` would overwrite the first instead of merging with it.

We don't necessarily need to support that syntax in this PR, but if we don't support it, we should at least display a useful error when the user tries to use it.

**Comment by rmartin16** [comment]
I meant to....but selecting "comment" instead probably would have been a better choice perhaps. I was mostly trying to communicate "don't feel required to send this back to me with changes".

**Review by rmartin16 — COMMENTED** [review]
updating status to comment

File: src/briefcase/commands/base.py
Comment (freakboy3742) [reply src/briefcase/commands/base.py:729]: Interesting - the very next option defined (`--verbosity`) *did* have a period. I've corrected both.

File: src/briefcase/commands/base.py
Comment (freakboy3742) [reply src/briefcase/commands/base.py:123]: Might as well be explicit.

File: src/briefcase/commands/base.py
Comment (freakboy3742) [reply src/briefcase/commands/base.py:119]: Good call. Right now, it's probably easier to reject multi-level keys completely. We can always add them later if there's a reasonable use case.

Comment (freakboy3742) [reply src/briefcase/commands/base.py:119]: Actually - in the process of looking at this, it occurred to me that this approach avoids validation of any provided value... and when I went looking to see how I could do validation, I found a *much* cleaner way to handle the overrides, in a way that both validates *and* avoids the issue @rmartin16 found with not passing the overrides to some commands.

**Comment by freakboy3742** [comment]
I've updated the PR so that any command-line overrides go through the same validation process as if they were specified in pyproject.toml. This ends up being a lot less invasive on commands, in addition to the validation benefit. 

I've also beefed up the validation that is done on the keys, ensuring that deep keys are rejected.

File: src/briefcase/commands/run.py
Comment (freakboy3742) [reply src/briefcase/commands/run.py]: Yes, but not any more with the revised approach that I've implemented.

File: src/briefcase/commands/update.py
Comment (freakboy3742) [reply src/briefcase/commands/update.py]: Again, Yes, but not any more with the revised approach that I've implemented.

**Comment by mhsmith** [comment]
@rmartin16: FYI, if you want to cancel an approval, you have to choose "Dismiss review" under "Changes approved" at the bottom of the page.

**Review by mhsmith — APPROVED** [review]

**Pull request merged**
